### ISSUE TITLE
Allow autogeneration of data sources

### ISSUE THREAD

**Issue by melinath** [hashicorp/terraform-provider-google#12595]
Data sources are currently all handwritten; however, they are fairly simple. We should allow autogeneration so that we can more cheaply support the existing data sources and also more easily add new ones.

**Comment by rileykarson** [comment]
Discussed in chat recently after @ScottSuarez brought this up. Context from there: most singular datasources (i.e. `data.google_compute_address` are of limited utility, and this has not proven the best use of our time to work on. Plural datasources (i.e. `data.google_compute_addresses`) would be slightly more useful in conventional configurations. 

The _most_ value comes from custom datasources generally, though, without corresponding resources.

**Comment by ScottSuarez** [comment]
My concern with datasources implemented manually is that sometimes they don't use the resource read or even schema. This leads to datasources becoming out of date with resource implementation. I would rather have flag to flip to turn on dataset generation if we want to keep it an opt-in thing. Thus ensuring consistency in implementation design and less foreseeable maintenance.

**Comment by rileykarson** [comment]
Yes, older datasources were handwritten. We avoid those now for the reasons you've mentioned and use the datasource wrapper for singular datasources. That's a separate topic than my comment which addresses why generating these has not been particularly high priority- singular datasources are not particularly valuable for end users in most cases.

> I would rather have flag to flip to turn on dataset generation if we want to keep it an opt-in thing.

Datasource?

**Comment by ScottSuarez** [comment]
mistype, mm gotcha !

**Comment by luckyswaminathan** [comment]
adding comment so I can be assigned

**Comment by c2thorn** [comment]
Made a more specific issue to finish out datasource gen in https://github.com/hashicorp/terraform-provider-google/issues/25987
